### PR TITLE
WildFly Standalone CentOS

### PR DESCRIPTION
**Pull request by danieloh30**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [v] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog
@rgardler 
* Create a new template for WildFly on CentOS on Azure
* This template is an alternative one for users who don't have Red Hat subscription but they want to explore Red Hat JBoss EAP features on Azure during the workshop.

### CODE REVIEW

**Comment by bmoore-msft** [comment]
@danieloh30 - for each of these where we are skipping CI I'll need instructions on how to do a manual deployment test - can email that to me if needed

**Comment by danieloh30** [comment]
@bmoore-msft In my case, I go to the page, "VM-Redhat - JBoss EAP 7 standalone mode" template[1] and click on "Deploy to Azure". After that, you can edit the existing template with this one. 
You should replace the "_artifactsLocation" parameter with "https://raw.githubusercontent.com/danieloh30/azure-quickstart-templates/wildfly-standalone-centos-1.0/wildfly-standalone-centos7/" before this PR is merged temporarily.

[1]https://github.com/danieloh30/azure-quickstart-templates/tree/wildfly-standalone-centos7-1.0/jboss-eap-standalone-rhel7

**Comment by bmoore-msft** [comment]
I need valid parameter values - or you need to add proper GEN values and remove .ci_skip

**Comment by danieloh30** [comment]
Here are sample parameters and you can input anything you want.

adminUsername: danieloh
adminPassword: Pa$$w0rd1!
dnsLabelPrefix: wildfly-doh
wildflyUserName: wildflyuser
wildflyPassword: redhat
sshPassPhrase: redhat

**Comment by bmoore-msft** [comment]
Thanks - but can you change those passwords and email the creds to me - so everyone doesn't have access to them?  (and do the same for the other PRs you have open).  My email alias is "bmoore".

**Comment by danieloh30** [comment]
@bmoore-msft Updated parameters to "GEN-UNIQUE" and "GEN-PASSWORD". Move forward to the next step.

File: wildfly-standalone-centos7/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        "description": "Linux VM user account name"
+      }
+    },
+    "adminPassword": {
+      "type": "securestring",
+      "metadata": {
+        "description": "Administrator password for Linux VM user account"
+      }
+    },
+    "dnsLabelPrefix": {
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: add a defaultValue using uniqueString()

File: wildfly-standalone-centos7/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    "storageAccountName": "[concat('vsts8',variables('baseName'))]",
+    "ScriptFolder": "scripts",
+    "ScriptFileName": "wildfly-setup-redhat.sh",
+    "subnet1Ref": "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('virtualNetworkName'), variables('subnet1Name'))]"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[variables('StorageAccountName')]",
+      "apiVersion": "2016-01-01",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: update apiVersion

File: wildfly-standalone-centos7/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+      "location": "[parameters('location')]",
+      "sku": {
+        "name": "[variables('storageAccountType')]"
+      },
+      "kind": "Storage",
+      "properties": {}
+    },
+    {
+      "apiVersion": "2017-10-01",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: update apiVersion

File: wildfly-standalone-centos7/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+      "location": "[parameters('location')]",
+      "properties": {
+        "publicIPAllocationMethod": "[variables('publicIPAddressType')]",
+        "dnsSettings": {
+          "domainNameLabel": "[parameters('dnsLabelPrefix')]"
+        }
+      }
+    },
+    {
+      "apiVersion": "2017-10-01",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: update apiVersion

File: wildfly-standalone-centos7/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+              "access": "Allow",
+              "priority": 1010,
+              "direction": "Inbound"
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion": "2017-10-01",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: update apiVersion

File: wildfly-standalone-centos7/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    "frontEndNSGName": "[concat(parameters('dnsLabelPrefix'),'-nsg8')]",
+    "storageAccountType": "Standard_LRS",
+    "storageAccountName": "[concat('vsts8',variables('baseName'))]",
+    "ScriptFolder": "scripts",
+    "ScriptFileName": "wildfly-setup-redhat.sh",
+    "subnet1Ref": "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('virtualNetworkName'), variables('subnet1Name'))]"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: how is this used?

File: wildfly-standalone-centos7/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+              "networkSecurityGroup": {
+                "id": "[resourceId('Microsoft.Network/networkSecurityGroups', variables('frontEndNSGName'))]"
+              }
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion": "2017-10-01",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: StorageAccountName

File: wildfly-standalone-centos7/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+              "subnet": {
+                "id": "[variables('subnet1Ref')]"
+              }
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion":  "2017-03-30",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: StorageAccountName

File: wildfly-standalone-centos7/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+              "id": "[resourceId('Microsoft.Network/networkInterfaces',variables('nicName'))]"
+            }
+          ]
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Compute/virtualMachines/extensions",
+      "name": "[concat(variables('vmName'),'/newuserscript')]",
+      "apiVersion": "2017-12-01",
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: StorageAccountName

File: wildfly-standalone-centos7/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        "[concat('Microsoft.Compute/virtualMachines/', variables('vmName'))]"
+      ],
+      "properties": {
+        "publisher": "Microsoft.Azure.Extensions",
+        "type": "CustomScript",
+        "typeHandlerVersion": "2.0",
+        "autoUpgradeMinorVersion": true,
+        "settings": {
+          "fileUris": [
+            "[concat(parameters('_artifactsLocation'), '/', variables('ScriptFolder'), '/', variables('ScriptFileName'), parameters('_artifactsLocationSasToken'))]"
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/azuredeploy.json]: use the uri function per guidelines

uri(parameters('_artifactsLocation'), concat(variables('ScriptFolder'), '/', variables('ScriptFileName'), parameters('_artifactsLocationSasToken')))

File: wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+export WILDFLY_USER=$2
+export WILDFLY_PASSWORD=$3
+
+echo "WILDFLY_USER: " ${WILDFLY_USER} >> /home/$1/install.log
+echo "WILDFLY_PASSWORD: " ${WILDFLY_PASSWORD} >> /home/$1/install.log
+
+echo "WILDFLY Downloading..." >> /home/$1/install.log
+cd /home/$1
+yum install -y git unzip java
+curl https://download.jboss.org/wildfly/16.0.0.Final/wildfly-16.0.0.Final.zip -o wildfly-16.0.0.Final.zip
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: Add some error handling and retries to the download

File: wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+unzip wildfly-16.0.0.Final.zip
+
+echo "Sample app deploy..." >> /home/$1/install.log 
+git clone https://github.com/danieloh30/dukes.git
```
Comment (bmoore-msft) [inline wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: you cannot clone a personal repo - you can add the files to the sample or use an official repo

File: wildfly-standalone-centos7/azuredeploy.json (hunk 2 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: Which version?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 3 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: Which version?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 4 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: Which version?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 5 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: Which version?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 8 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: What does StorageAccountName mean?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 9 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: What does StorageAccountName mean?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 10 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: What guidelines? Is the URI function necessary? The other templates have already implemented using the above function.

File: wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh (hunk 11 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: Provide the guidelines.

File: wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh (hunk 12 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: It doesn't make sense. There are templates to use a private git repo. For example, https://github.com/danieloh30/azure-quickstart-templates/blob/master/jboss-eap-standalone-rhel7/scripts/eap-setup-redhat.sh

What is the official repo? How do you define it?

Comment (bmoore-msft) [reply wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: A non-personal repo - if there isn't one, add the files to this sample

File: wildfly-standalone-centos7/azuredeploy.json (hunk 2 above)
Comment (bmoore-msft) [reply wildfly-standalone-centos7/azuredeploy.json]: See the BP guidelines for the repo

File: wildfly-standalone-centos7/azuredeploy.json (hunk 10 above)
Comment (bmoore-msft) [reply wildfly-standalone-centos7/azuredeploy.json]: The BP guidelines for submitting samples to this repo

File: wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh (hunk 12 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: Check out the existing one.

https://github.com/Azure/azure-quickstart-templates/blob/master/jboss-eap-standalone-rhel7/scripts/eap-setup-redhat.sh#L42

File: wildfly-standalone-centos7/azuredeploy.json (hunk 10 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: Check this out:

"properties": {
   "publisher": "Microsoft.Azure.Extensions",
   "type": "CustomScript",
   "version": "2.0",
   "autoUpgradeMinorVersion": true,
  "settings": {
    "fileUris": [
      "[concat(variables('template').assets, '/lamp-app/install_lamp.sh')]"
      ]
  },
  "protectedSettings": {
      "commandToExecute": "[concat('sh install_lamp.sh ', parameters('mySqlPassword'))]"
  }
}

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.

**Comment by danieloh30** [comment]
@bmoore-msft What do you mean "Closing abandoned request - please reopen if you can address the feedback."?  Do you want me to create another PR after I closed this? You should check out other templates that are developed in the same way.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 2 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: There is no specific version for this. Let me know the exact version that you need to update.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 7 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: Which version?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 6 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: What do you mean? storageAccounts is necessary to create a VM.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 2 above)
Comment (bmoore-msft) [reply wildfly-standalone-centos7/azuredeploy.json]: Read the contribution guide: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#api-versions

File: wildfly-standalone-centos7/azuredeploy.json (hunk 6 above)
Comment (bmoore-msft) [reply wildfly-standalone-centos7/azuredeploy.json]: storageAccounts are not necessary for a VM and you're not using it for the VM.

**Comment by bmoore-msft** [comment]
The are a number of comments on this PR, those comments need to be addressed before the PR can be merged.

**Comment by danieloh30** [comment]
> The are a number of comments on this PR, those comments need to be addressed before the PR can be merged.
@bmoore-msft 
You should provide a more specific example of how to modify it rather than just following the guide. For example, which API version would be the best suitable for updating? If not, you force me to test all APIs that created less than 6 months ago. I need a more productive and effective way to improve the codes. You should go through existing templates of RH JBoss as well. Please understand the template and JBoss first and update more specific guidelines.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 1 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft Why should the parameter be unique? If so, what is your recommendation?

File: wildfly-standalone-centos7/azuredeploy.json (hunk 2 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft According to the the link, I can use the latest one or no more than 12 months old. Too vague, I need a certain version for this.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 3 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft According to the the link, I can use the latest one or no more than 12 months old. Too vague, I need a certain version for this.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 4 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft According to the link, I can use the latest one or no more than 12 months old. Too vague, I need a certain version for this.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 5 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft According to the link, I can use the latest one or no more than 12 months old. Too vague, I need a certain version for this.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 8 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft According to the link, I can use the latest one or no more than 12 months old. Too vague, I need a certain version for this.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 9 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft According to the link, I can use the latest one or no more than 12 months old. Too vague, I need a certain version for this.

File: wildfly-standalone-centos7/azuredeploy.json (hunk 10 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/azuredeploy.json]: @bmoore-msft Have you check the above example that I wrote?

File: wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh (hunk 11 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: @bmoore-msft Any updates?

File: wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh (hunk 12 above)
Comment (danieloh30) [reply wildfly-standalone-centos7/scripts/wildfly-setup-redhat.sh]: @bmoore-msft Any comments?

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.

**Pull request closed without merging**